File: sdk-go/iwf/command_results.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package iwf

import "github.com/superdurable/iwf/sdk-go/gen/iwfidl"

type (
	CommandResults struct {
		Timers                  []TimerCommandResult
		Signals                 []SignalCommandResult
		InternalChannelCommands []InternalChannelCommandResult
		WaitUntilApiSucceeded   *bool
	}

	SignalCommandResult struct {
		CommandId   string
		ChannelName string
		SignalValue Object
		Status      iwfidl.ChannelRequestStatus
	}

	TimerCommandResult struct {
		CommandId string
		Status    iwfidl.TimerStatus
	}

	InternalChannelCommandResult struct {
		CommandId   string
		ChannelName string
		Value       Object
		Status      iwfidl.ChannelRequestStatus
	}
)

func (c CommandResults) GetTimerCommandResultById(id string) *TimerCommandResult {
	for _, cmd := range c.Timers {
		if cmd.CommandId == id {
			return &cmd
		}
	}
	return nil
}

func (c CommandResults) GetSignalCommandResultById(id string) *SignalCommandResult {
	for _, cmd := range c.Signals {
		if cmd.CommandId == id {
			return &cmd
		}
	}
	return nil
}

func (c CommandResults) GetInternalChannelCommandResultById(id string) *InternalChannelCommandResult {
	for _, cmd := range c.InternalChannelCommands {
		if cmd.CommandId == id {
			return &cmd
		}
	}
	return nil
}

func (c CommandResults) GetSignalCommandResultByChannel(channelName string) *SignalCommandResult {
	for _, cmd := range c.Signals {
		if cmd.ChannelName == channelName {
			return &cmd
		}
	}
	return nil
}

func (c CommandResults) GetInternalChannelCommandResultByChannel(channelName string) *InternalChannelCommandResult {
	for _, cmd := range c.InternalChannelCommands {
		if cmd.ChannelName == channelName {
			return &cmd
		}
	}
	return nil
}

func (c CommandResults) GetWaitUntilApiSucceeded() *bool {
	return c.WaitUntilApiSucceeded
}

File: sdk-java/src/main/java/io/iworkflow/core/command/CommandResults.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.iworkflow.core.command;

import io.iworkflow.core.communication.InternalChannelCommandResult;
import io.iworkflow.core.communication.SignalCommandResult;
import org.immutables.value.Value;

import java.util.List;
import java.util.Optional;

/**
 * This is the container of all requested commands' results/statuses
 */
@Value.Immutable
public abstract class CommandResults {

    //public abstract List<LongRunningActivityCommandResult> getAllLongRunningActivityCommandResults();
    public abstract List<SignalCommandResult> getAllSignalCommandResults();

    public abstract List<TimerCommandResult> getAllTimerCommandResults();

    public abstract List<InternalChannelCommandResult> getAllInternalChannelCommandResult();

    public abstract Optional<Boolean> getWaitUntilApiSucceeded();

    // below are helpers
    public <T> T getSignalValueByIndex(int idx) {
        final List<SignalCommandResult> results = getAllSignalCommandResults();
        final SignalCommandResult value = results.get(idx);
        return (T) value.getSignalValue().get();
    }

    public <T> T getSignalValueById(String commandId) {
        final List<SignalCommandResult> results = getAllSignalCommandResults();
        for (SignalCommandResult result : results) {
            if (result.getCommandId().equals(commandId)) {
                return (T) result.getSignalValue().get();
            }
        }
        throw new IllegalArgumentException("commandId not found");
    }
}
